### DISCUSSION TITLE
OAuth2 authentication using route handlers in Nextjs13

### DISCUSSION THREAD

**Discussion by gaganbiswas** [pocketbase/pocketbase#4349] [Q&A] [answered]
```js
import open from 'open';
import {pb} from '@/lib/pocketbase.ts'

export const GET = async (req: Request) => {
try {
    const result = await pb.client.collection("users").authWithOAuth2({
      provider: "google",
      urlCallback: async (url) => {
        await open(url);
      },
    });

    if (!result?.token)
      return NextResponse.json({
        status: 0,
        message: "Error processing your request.",
      });

    const { record, token } = result;
    record.token = token;
    cookies().set("pb_auth", pb.client.authStore.exportToCookie(), {
      httpOnly: true,
    });
    return NextResponse.json({
      status: 1,
      message: `Welcome back, ${record.name}`,
    });
  } catch (error: any) {
    console.log(error);
    return NextResponse.json({
      status: 0,
      message: error.message,
    });
  }
}
```
This is my current implementation of the Oauth2 in the route handlers and it works fine in opening the window and signing in the user and storing the cookie and all.
I currently am facing two problems with this approach.

1. The window that is being opened is in a different tab which doesn't automatically closes (bad user experience)
2. If the user closes the tab/window, I don't get any response back, which is because urlCallback doesn't return anything.

**Comment by ganigeorgiev** [comment, marked as answer]
Are you sure that the above is functional?

Unless Next.js has some special handling for the "open" dependency (which I assume is https://www.npmjs.com/package/open) I don't understand how the above is working in the first place.

The `authWithOAuth2` helper starts a realtime connection and needs to open a browser window on the same client device (you can't just open remotely new windows/tabs for the user from a plain HTTP server response unless you combine it with browser JS but I don't think that's what happens here and probably it works only on your localhost as node spawns a new local xdg or other similar process).

If you want to handle OAuth2 authentication on the server-side you'll need the "Manual code exchange" flow like in the second tab of the [OAuth2 integration](https://pocketbase.io/docs/authentication/#oauth2-integration) guide. The example in the docs is using plain browser JS but the idea is the same - you'll need a page to show the "Login with ..." links and another route that will handle the OAuth2 redirect and will call `authWithOAuth2Code()` (aka. the actual authentication that will exchange the OAuth2 code for token).

**Reply by ganigeorgiev** [reply]
Also `cookies().set("pb_auth", ...)` I think expect the second argument to be the plain value, while the SDK `exportToCookie()` returns a raw cookie string, aka. `pb_auth=...`. If you are using the Next.js cookies abstraction you don't need the SDK cookie helpers and can just assign the `pb.authStore.token` as value.

**Reply by gaganbiswas** [reply]
Okay. I just tried it in a preview environment for testing and it is not working. Looks like this only works in my local development env.
So, I will just then integrate the `authWithOAuth2Code()`. Thank you for the help!

**Reply by gaganbiswas** [reply]
Hey, I was having a lot of problem in setting this OAuth. I went back to default setup (client side) and now I am facing this CORS error `Cross-Origin Request Blocked` only when trying to authenticate using OAuth. Other stuffs like fetching the data or other CRUD operations are working fine. 

`pocketbase.ts`
```js
import { ReadonlyRequestCookies } from "next/dist/server/web/spec-extension/adapters/request-cookies";
import PocketBase from "pocketbase";

export const pb = new PocketBase(process.env.NEXT_PUBLIC_POCKETBASE_URL);
pb.autoCancellation(false);

export const isAuthenticated = async (cookieStore: ReadonlyRequestCookies) => {
  const cookie = cookieStore.get("pb_auth");
  if (!cookie) {
    return false;
  }

  pb.authStore.loadFromCookie(cookie?.value || "");
  return pb.authStore.isValid || false;
};
```

`google-button.tsx`
```js
"use client";

import React, { useState } from "react";
import { Button } from "@/components/ui/button";
import { toast } from "sonner";
import { pb } from "@/lib/pocketbase";

const GoogleButton = () => {
  const [loading, setLoading] = useState(false);

  const handleGoogleLogin = async () => {
    setLoading(true);

    try {
      const authData = await pb.collection("users").authWithOAuth2({
        provider: "google",
      });

      console.log(authData);
    } catch (error: any) {
      toast.error(error.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <Button
      type="button"
      variant="outline"
      className="border border-slate-500 text-slate-900 hover:text-slate-900"
      size={"lg"}
      disabled={loading}
      onClick={handleGoogleLogin}
    >
      <GoogleIcon className="mr-2 h-6 w-6" />
      Continue with Google
    </Button>
  );
};

export default GoogleButton;
```

**Reply by gaganbiswas** [reply]
Also, in I have tried this in production not from my localhost and I get the same error there as well.

Edited: Only happening when trying to access resources from the frontend (client-side)

Edited: Fixed with manual code exchange.
